metrics/prommetrics.py

```diff
@@ -23,43 +29,35 @@ def connect(self):
 
     def push_metrics(self, observer):
-        self._get_metrics_from_prometheus(observer) # Initial run for metric collection
-
-        # Scheduler schedules a background job that needs to be run regularly
-        scheduler = BackgroundScheduler()
-        scheduler.start()
-        scheduler.add_job(
-            func=lambda: self._get_metrics_from_prometheus(observer), # Run this function every 5 minutes to poll for new metric data
-            trigger=IntervalTrigger(minutes=5),
-            id='update_metric_data',
-            name='Ticker to get new data from prometheus',
-            replace_existing=True)
-
-        atexit.register(lambda: scheduler.shutdown()) # Shut down the scheduler when exiting the app
+        self.observe_prom_metrics_range(observer=observer,
+                                        metrics_list=self.metrics_list,
+                                        start_time=self.metric_start_datetime,
+                                        end_time=self.metric_end_datetime,
+                                        chunk_size=self.metric_chunk_size)
 
-    def _get_metrics_from_prometheus(self, observer=None):
+    def observe_prom_metrics_range(self, observer, metrics_list, start_time, end_time='now', chunk_size='1h'):
         # Collect credentials to connect to a prometheus instance
         prom_token = os.getenv("PROM_ACCESS_TOKEN")
         prom_url = os.getenv("PROM_URL")
         if not (prom_token or prom_url):
             sys.exit("Error: Prometheus credentials not found")
-
-
-        prom = Prometheus(url=prom_url, token=prom_token, data_chunk='5m',stored_data='5m')
-
-        metrics_list = prom.all_metrics() # Get a list of all the metrics available from Prometheus
-
-        print("Polling Prometheus for new metric data")
-
-        metric_data = dict()
-        if observer:
-            for metric in metrics_list:
-                pkt = (json.loads(prom.get_metric(name=metric))[0])
-                metric_data[metric] = pkt
-                observer.on_next(pkt) # push metric data to the Observer
-            pass
-        else:
-            for metric in metrics_list:
-                metric_data[metric] = (json.loads(prom.get_metric(name=metric))[0])
-
-        return(metric_data)
+        prom = Prometheus(url=prom_url, token=prom_token)
+
+        # Calculate chunk size to download and push to the observer at each instance
+        chunk_seconds = int(round((dateparser.parse('now') - dateparser.parse(chunk_size)).total_seconds()))
+
+        start = dateparser.parse(start_time).timestamp()
+        end = dateparser.parse(end_time).timestamp()
+
+        while start <= end:     # Main loop which iterates through time-ranges to collect a chunk of data at every iteration
+            for metric_name in metrics_list:    # Loop to get a chunk of data for every metric in the list
+                pkt_list = json.loads(prom.get_metric_range_data(metric_name=metric_name, start_time=start, end_time=end))
```

Comment:
can't we get rid of `json.loads` and `json.dumps` ?
How about `prometheus.py` always returning a `json.loads` dict?